clusterloader2/pkg/measurement/common/wait_for_jobs.go

```diff
@@ -0,0 +1,247 @@
+	w.clusterFramework = config.ClusterFramework
+
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		if err = w.selector.Parse(config.Params); err != nil {
+			return nil, err
+		}
+		return nil, w.start()
+	case "gather":
+		timeout, err := util.GetDurationOrDefault(config.Params, "timeout", defaultWaitForCompletedJobsTimeout)
+		if err != nil {
+			return nil, err
+		}
+		return nil, w.gather(timeout)
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func (w *waitForCompletedJobsMeasurement) Dispose() {
+	if !w.isRunning {
+		return
+	}
+	w.isRunning = false
+	w.queue.Stop()
+	w.cancel()
+}
+
+func (w *waitForCompletedJobsMeasurement) String() string {
+	return waitForCompletedJobsName
+}
+
+func (w *waitForCompletedJobsMeasurement) start() error {
+	if w.isRunning {
+		klog.V(2).Infof("%v: wait for completed jobs measurement already running", w)
+		return nil
+	}
+	klog.V(2).Infof("%v: starting wait for completed jobs measurement...", w)
+	w.isRunning = true
+	ctx, cancel := context.WithCancel(context.Background())
+	w.cancel = cancel
+	c := w.clusterFramework.GetClientSets().GetClient()
+	inf := informer.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				w.selector.ApplySelectors(&options)
+				return c.BatchV1().Jobs(w.selector.Namespace).List(ctx, options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				w.selector.ApplySelectors(&options)
+				return c.BatchV1().Jobs(w.selector.Namespace).Watch(ctx, options)
+			},
+		},
+		func(oldObj, newObj interface{}) {
+			f := func() {
+				w.handleObject(oldObj, newObj)
+			}
+			w.queue.Add(&f)
+		},
+	)
+	return informer.StartAndSync(inf, ctx.Done(), informerSyncTimeout)
+}
+
+func (w *waitForCompletedJobsMeasurement) gather(timeout time.Duration) error {
+	if !w.isRunning {
+		return fmt.Errorf("%v: wait for completed jobs was not started", w)
+	}
+	klog.V(2).Infof("%v: waiting for completed jobs measurment...", w)
+	jobKeys, err := w.objectKeys()
+	if err != nil {
+		return err
+	}
+
+	cond := func() (bool, error) {
+		w.lock.Lock()
+		defer w.lock.Unlock()
+		finishedKeys := make(sets.String, len(w.finishedJobs))
+		for k := range w.finishedJobs {
+			finishedKeys.Insert(k)
+		}
+		return jobKeys.Equal(finishedKeys), nil
+	}
+	if err := wait.Poll(checkCompletedJobsInterval, timeout, cond); err != nil {
+		klog.V(2).Infof("Timed out waiting for all jobs to complete: %w", err)
+	}
+	w.lock.Lock()
+	completed := 0
+	failed := 0
+	for _, cond := range w.finishedJobs {
+		if cond == batchv1.JobComplete {
+			completed++
+		} else if cond == batchv1.JobFailed {
+			failed++
+		}
+	}
+	w.lock.Unlock()
+	klog.V(2).Infof("%v: %d/%d Jobs finished, %d completed, %d failed", w, completed+failed, len(jobKeys), completed, failed)
+	return nil
```

Comment:
Do we want to return error if some of jobs failed? 
I am wondering how exactly we want to use this measurement later. 
@jprzychodzen @alculquicondor

Reply:
What are the implications of error? Does the framework continue with the next steps? I noticed that `WaitForControlledPodsRunning` doesn't trigger an error if all pods are not running (unless I misread).

We weren't planning purposely failing jobs for now, but we could consider in the future.

For the tests we are planning, a failed job should be unexpected. But if there was an error, it would be helpful to complete other measurements, such as gathering metrics.

Reply:
Yes, it continues with next steps even if this measurement fails. Here is example when `WaitForControlledPodsRunning` returns error when some of deployments time out https://github.com/kubernetes/perf-tests/blob/c50dea13450a0c3a6aeee677124da8a832665c88/clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go#L270

Reply:
Ah, missed it. For now, it makes sense to error when some jobs didn't finish (regardless of their success)